### PR TITLE
Refactor print statements to log statements and internationalize messages

### LINKED ISSUE #49: Identify all print statements that don't use log4j and refactor them
The following kinds of print statements need to be refactored as mentioned:
```
System.out.println(...) -> log.info(...) or log.debug(...)
System.err.println(...) -> log.error(...)
e.printStackTrace() -> log.error(e)
```
Internationalize all printed/logged statements, including exception messages.

### PR DESCRIPTION
**Pull request by silvs110**
closes #49.
- Refactor print statements to log statements.
- Internationalize log messages.
- Rethrow logged exceptions.

### CODE REVIEW

File: src/test/java/ShellUtilityTest.java [hunk 1]
```diff
@@ -44,7 +48,7 @@ void prepare() {
 
     @Test
     public void testExecuteCommandString() throws IOException, InterruptedException, TimeoutException, ShellUtility.ShellNotFoundException {
-        System.out.println("SHELL - " + typeOfShell);
+        logger.info("SHELL - " + typeOfShell);
```
Comment (the-c0d3br34k3r) [inline src/test/java/ShellUtilityTest.java:51]: Not needed. Can be removed.

File: src/test/java/RegexUtilityTest.java [hunk 2]
```diff
@@ -12,15 +15,21 @@
...
+        String patternString="\\d+\\s+\\S+";
+        String stringInput="1290 TestString";
         Assertions.assertTrue(
             RegexUtility.matches(
-                    "\\d+\\s+\\S+",
+                    patternString,
                     0,
-                    "1290 TestString"),
-            "Regex should've matched the provided string."
+                    stringInput),
```
Comment (the-c0d3br34k3r) [inline src/test/java/RegexUtilityTest.java:27]:
```suggestion
                    stringInput
            ),
```

File: src/main/java/datastructures/tree/Tree.java [hunk 3]
```diff
@@ -44,7 +50,9 @@ public void addChildNode(T child) {
 			tree_changed=true;
 		}
 		else {
-			System.err.println("Node Capacity Full, Cannot have more children!");
+			logger.error(
+					I18NUtility.getString("datastructures.Tree.maxNumberOfChildrenNodesExceededMessage")
```
Comment (the-c0d3br34k3r) [inline src/main/java/datastructures/tree/Tree.java:54]:
```suggestion
		I18NUtility.getString("datastructures.Tree.maxNumberOfChildrenNodesExceededMessage")
```

**Review by the-c0d3br34k3r — COMMENTED** [review]
@silvs110 Can you whip up a quick unit test that fails whenever it detects a System.out.println/System.err.println/.printStackTrace() in the code?

File: src/main/java/datastructures/tree/AVLTree.java [hunk 4]
```diff
@@ -25,10 +36,12 @@ public void delete(T node_value) {
...
 	public void addChildNode(T child_value) {
 		if(getInsertionIndex(child_value)==-1)
-			System.out.println("Value already present.");
+			logger.debug(
+					I18NUtility.getString("datastructures.valueAlreadyPresentMessage")
+			);
 		else {
 			sorted_node_values.add(getInsertionIndex(child_value), child_value);
-			System.out.println("Sorted List - "+sorted_node_values.toString());
+			logger.debug("Sorted List - "+ sorted_node_values.toString());
```
Comment (the-c0d3br34k3r) [inline src/main/java/datastructures/tree/AVLTree.java:44]: Can be removed.

File: src/test/java/ShellUtilityTest.java (hunk 1 above)
Comment (silvs110) [reply src/test/java/ShellUtilityTest.java:51]: done

File: src/test/java/RegexUtilityTest.java (hunk 2 above)
Comment (silvs110) [reply src/test/java/RegexUtilityTest.java:27]: done

File: src/main/java/datastructures/tree/AVLTree.java (hunk 4 above)
Comment (silvs110) [reply src/main/java/datastructures/tree/AVLTree.java:44]: done

**Comment by silvs110** [comment]
> @silvs110 Can you whip up a quick unit test that fails whenever it detects a System.out.println/System.err.println/.printStackTrace() in the code?

done

File: src/main/java/datastructures/list/DoublyLinkedList.java [hunk 5]
```diff
@@ -1,12 +1,25 @@
...
+import org.apache.logging.log4j.Logger;
+import utilities.I18NUtility;
+
 public class DoublyLinkedList<T> extends LinkedList<T> {
+
+	private final Logger logger = LogManager.getLogger(DoublyLinkedList.class);
+
 	public DoublyLinkedList() {
 		// TODO Auto-generated constructor stub
+
```
Comment (the-c0d3br34k3r) [inline src/main/java/datastructures/list/DoublyLinkedList.java:13]: Why is there a new line here?

File: src/main/java/datastructures/list/DoublyLinkedList.java [hunk 6]
```diff
@@ -1,12 +1,25 @@
...
 	}
 	public void removeNextNode(Node<T> temp) {
 		Node<T> temp1;
-		System.out.println("Removing node after "+temp.value);
+		logger.debug(
+				I18NUtility.getFormattedString(
+						"datastructures.Node.removeNodeInfo",
+						temp.value
+				)
```
Comment (the-c0d3br34k3r) [inline src/main/java/datastructures/list/DoublyLinkedList.java:21]:
```suggestion
				I18NUtility.getFormattedString(
					"datastructures.Node.removeNodeInfo",
					temp.value
				)
```

File: src/main/java/datastructures/list/LinkedList.java [hunk 7]
```diff
@@ -16,7 +22,9 @@ public T getNodeValue(int index) {
 				return temp.getValue();
 			temp = temp.getNextNode();
 		}
-		System.err.println("Invalid Index");
+		logger.error(
+				I18NUtility.getFormattedString("datastructures.invalidIndexBasicErrorMessage")
```
Comment (the-c0d3br34k3r) [inline src/main/java/datastructures/list/LinkedList.java:26]: Alignment issue.

File: src/main/java/datastructures/list/LinkedList.java [hunk 8]
```diff
@@ -30,7 +38,9 @@ public T getNodeValue(int index) {
 				return temp;
 			temp = temp.getNextNode();
 		}
-		System.err.println("Invalid Index");
+		logger.error(
+				I18NUtility.getFormattedString("datastructures.invalidIndexBasicErrorMessage")
```
Comment (the-c0d3br34k3r) [inline src/main/java/datastructures/list/LinkedList.java:42]: Alignment issue.

File: src/main/java/datastructures/list/LinkedList.java [hunk 9]
```diff
@@ -51,7 +61,9 @@ else if(i==index-1) {
 			}
 		}
 		else {
-			System.err.println("Invalid index for insertion.");
+			logger.error(
+					I18NUtility.getString("datastructures.invalidIndexInsertionErrorMessage")
```
Comment (the-c0d3br34k3r) [inline src/main/java/datastructures/list/LinkedList.java:65]: Alignment issue.

File: src/main/java/datastructures/list/LinkedList.java [hunk 10]
```diff
@@ -108,7 +120,9 @@ public void removeIndex(int index) {
 			}
 		}
 		else {
-			System.err.println("Invalid index for Deletion.");
+			logger.error(
+					I18NUtility.getString("datastructures.invalidIndexDeletionErrorMessage")
```
Comment (the-c0d3br34k3r) [inline src/main/java/datastructures/list/LinkedList.java:124]: Alignment issue.

File: src/main/java/datastructures/queue/PriorityQueue.java [hunk 11]
```diff
@@ -48,7 +54,9 @@ else if(i==index-1) {
 			}
 		}
 		else {
-			System.err.println("Invalid index for insertion.");
+			logger.error(
+					I18NUtility.getString("datastructures.invalidIndexInsertionErrorMessage")
```
Comment (the-c0d3br34k3r) [inline src/main/java/datastructures/queue/PriorityQueue.java:58]: Alignment issue.

File: src/main/java/datastructures/tree/AVLTree.java [hunk 12]
```diff
@@ -16,7 +22,12 @@ public AVLTree(T value) {
 	
 	public void delete(T node_value) {  
 		if(find(node_value)==-1)
-			System.out.println("Value to be deleted not found.");
+			logger.error(
+					I18NUtility.getFormattedString(
```
Comment (the-c0d3br34k3r) [inline src/main/java/datastructures/tree/AVLTree.java:26]: Alignment issue.

File: src/test/java/CodingPracticesTest.java [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+     * Test that all Java files in the project don't have print statements.
+     */
+    @Test
+    void testJavaFilesForPrintStatements(){
+        javaFilesPath
+                .forEach(javaFilePath -> {
+                    try {
+                        byte[] encoded = Files.readAllBytes(javaFilePath);
+                        String javaFileContent = new String(encoded, StandardCharsets.US_ASCII);
+                        Pattern pattern = Pattern.compile("(System\\.(out|err)\\.print|e\\.printStackTrace\\(\\))");
```
Comment (the-c0d3br34k3r) [inline src/test/java/CodingPracticesTest.java:48]: Please store the regular expression in a property file.

File: src/test/java/CodingPracticesTest.java [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+     * Get all Java Files in basePathString.
+     * @return the list of paths to the Java Files.
+     */
+    private static List<Path> getJavaFilesPath(){
+        Path basePath = Paths.get(basePathString).toAbsolutePath();
+        try {
+            return Files.walk(basePath)
+                    .filter(path -> (path.endsWith(".java") && Files.isRegularFile(path)))
+                    .collect(Collectors.toList());
+        }catch (IOException e) {
```
Comment (the-c0d3br34k3r) [inline src/test/java/CodingPracticesTest.java:32]:
```suggestion
        } catch (IOException e) {
```

File: src/test/java/CodingPracticesTest.java [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+        }
+    }
+
+    /**
+     * Test that all Java files in the project don't have print statements.
+     */
+    @Test
+    void testJavaFilesForPrintStatements(){
+        javaFilesPath
+                .forEach(javaFilePath -> {
```
Comment (the-c0d3br34k3r) [inline src/test/java/CodingPracticesTest.java:44]:
```suggestion
        javaFilesPath.forEach(javaFilePath -> {
```

File: src/test/java/CodingPracticesTest.java [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+                    try {
+                        byte[] encoded = Files.readAllBytes(javaFilePath);
+                        String javaFileContent = new String(encoded, StandardCharsets.US_ASCII);
```
Comment (the-c0d3br34k3r) [inline src/test/java/CodingPracticesTest.java:47]: Are they files encoded with ASCII or UTF-8?

File: src/test/java/CodingPracticesTest.java [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+                        Pattern pattern = Pattern.compile("(System\\.(out|err)\\.print|e\\.printStackTrace\\(\\))");
+                        Matcher matcher = pattern.matcher(javaFileContent);
+                        Assertions.assertFalse(
```
Comment (the-c0d3br34k3r) [inline src/test/java/CodingPracticesTest.java:50]: Please print the output as follows:

```
 The following files output to STDOUT/STDERR directly. Please change them to use the logger.

 a/b/c/d.java
     Line 26    |    System.out.println("Hello");

 a/b/c/e.java
     Line 29    |    System.out.println("Hola");
     Line 112   |    e.printStackTrace();
```

File: src/test/java/CodingPracticesTest.java [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+
+    /**
+     * Test that all Java files in the project don't have print statements.
+     */
+    @Test
+    void testJavaFilesForPrintStatements(){
+        javaFilesPath
+                .forEach(javaFilePath -> {
+                    try {
+                        byte[] encoded = Files.readAllBytes(javaFilePath);
```
Comment (the-c0d3br34k3r) [inline src/test/java/CodingPracticesTest.java:46]: Use this utility function:
```
    private static HashMap<Long, String> getStdOutErrLines(Path path) throws IOException {
        Stream<String> lines = Files.lines(path);
        Iterator<String> linesIterator = lines.iterator();

        HashMap<Long, String> stdOutErrLines = new HashMap<>();

        LongStream.rangeClosed(1, lines.count())
            .parallel()
            .forEach(lineNumber -> {
                String line = linesIterator.next();
                Matcher matcher = Pattern.compile("(System\\.(out|err)\\.print|e\\.printStackTrace\\(\\))") // Add regex to property file
                        .matcher(line);
                if(matcher.find()) {
                    stdOutErrLines.put(lineNumber, line);
                }
            });

        return stdOutErrLines;
    }
```

**Review by the-c0d3br34k3r — CHANGES_REQUESTED** [review]
Reviewed. Requested changes.

File: src/test/java/CodingPracticesTest.java (hunk 13 above)
Comment (silvs110) [reply src/test/java/CodingPracticesTest.java:48]: done

File: src/test/java/CodingPracticesTest.java (hunk 15 above)
Comment (silvs110) [reply src/test/java/CodingPracticesTest.java:44]: done

File: src/test/java/CodingPracticesTest.java (hunk 14 above)
Comment (silvs110) [reply src/test/java/CodingPracticesTest.java:32]: done

File: src/test/java/CodingPracticesTest.java (hunk 17 above)
Comment (silvs110) [reply src/test/java/CodingPracticesTest.java:50]: done

File: src/test/java/CodingPracticesTest.java (hunk 18 above)
Comment (silvs110) [reply src/test/java/CodingPracticesTest.java:46]: done

File: src/test/java/CodingPracticesTest.java [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+            throw new RuntimeException(e);
+        }
+        return propertyFile;
+    }
+
+    /**
+     * Get a mapping of the lines where print statements were found for a specific Java file.
+     * @param path the path of the Java file to look for print statements.
+     * @return A map of the lines where print statements were found.
+     * @throws IOException if an I/O exception occurs.
```
Comment (the-c0d3br34k3r) [inline src/test/java/CodingPracticesTest.java:68]: Seriously? xD
```suggestion
     * @throws IOException if there is an issue reading the file.
```

File: src/test/java/CodingPracticesTest.java [hunk 20]
```diff
@@ -0,0 +1,136 @@
...
+                        );
+                    }
+
+                } catch (IOException e) {
+                    logger.error(e);
+                    throw new RuntimeException(e);
+                }
+
+            });
+        String error_message = filesPrintStatements.entrySet()
```
Comment (the-c0d3br34k3r) [inline src/test/java/CodingPracticesTest.java:112]:
```suggestion
        String errorMessage = filesPrintStatements.entrySet()
```

**Review by the-c0d3br34k3r — CHANGES_REQUESTED** [review]
Reviewed, requested changes.

File: src/test/java/CodingPracticesTest.java (hunk 19 above)
Comment (silvs110) [reply src/test/java/CodingPracticesTest.java:68]: Laziness at its finest

File: src/test/java/SorterTest.java [hunk 21]
```diff
@@ -16,7 +17,11 @@ void testArraySorting(Integer[] inputArray, Integer[] expectedOutputArray) {
             Integer[] actualOutputArray = sorter.sort(inputArrayDupe, sortingTechnique);
             Assertions.assertTrue(
                     Arrays.equals(expectedOutputArray, actualOutputArray),
-                    sortingTechnique + " failed.\nResult: " + Arrays.toString(actualOutputArray)
+                    I18NUtility.getFormattedString(
```
Comment (the-c0d3br34k3r) [inline src/test/java/SorterTest.java:20]: Will this work?

**Review by the-c0d3br34k3r — COMMENTED** [review]
Requested clarifications.

File: src/test/java/CodingPracticesTest.java (hunk 19 above)
Comment (the-c0d3br34k3r) [reply src/test/java/CodingPracticesTest.java]: This is a new low for laziness, not a win.

File: src/test/java/SorterTest.java (hunk 21 above)
Comment (the-c0d3br34k3r) [reply src/test/java/SorterTest.java:20]: Looks like it will.

**Pull request merged**